### PR TITLE
ERC-2126: Signature Type Recognition

### PR DESCRIPTION
**Pull request by pedrouid**
Enable Dapps to be able to recognize how a given signature was signed (e.g. by owner of a smart contract, EOA, etc) so that they can use the correct mechanism to verify the signature.

Inspired by 0x Protocol V2.0 Signature Types

### CODE REVIEW

File: EIPS/2126.md [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ---
 eip: 2126
 title: Signature Type Recognition
-author: Pedro Gomes <pedro@pedrouid.com>, Ramesh Nair <ram@hiddentao.com>
+author: Pedro Gomes <@pedrouid>, Ramesh Nair <@hiddentao>
```
Comment (axic) [inline EIPS/2126.md]: Has to be parenthesis for github usernames.

File: EIPS/eip-2126.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+A real-world example would be using Metamask or [WalletConnect](https://walletconnect.org), where you could expose accounts which are either EOA accounts or smart contract accounts, which in turn would provide similar signatures but require different validation processes.
+
+The solution to this issue is to append a byte to the end of the signature which will correspond to one of the given signature types listed below. Dapps can then query this byte to know how to validate the given signature.
+
+_Note: this approach is [already in production use by 0x](https://github.com/0xProject/0x-protocol-specification/blob/master/v2/v2-specification.md#signature-types), and indeed this spec is based on their work._
+
+## Formatting Signatures
```
Comment (axic) [inline EIPS/eip-2126.md]: I think all these should go within a Specification section.

File: EIPS/eip-2126.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+The is the default value of the signature byte. A transaction that includes an Illegal signature will be reverted. Therefore, users must explicitly specify a valid signature type.
+
+### Invalid
+
+An `Invalid` signature always returns false. An invalid signature can always be recreated and is therefore offered explicitly. This signature type is largely used for testing purposes.
+
+### EIP712
+
+An [`EIP712`](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md) signature is considered valid if the address recovered from calling ecrecover with the given hash and decoded `v`, `r`, `s` values is the same as the specified signer. In this case, the signature is encoded in the following way:
```
Comment (axic) [inline EIPS/eip-2126.md]: Please use the "canonical URL", e.g. eips.ethereum.org/EIPS/eip-712

File: EIPS/eip-2126.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    "...data...",
+    "0x...",
+    "password",
+    null, // callback
+    {
+        includeType: true, // informs signer to include signature type
+    }
+})
+```
+
```
Comment (axic) [inline EIPS/eip-2126.md]: Missing the Rationale section.

File: EIPS/eip-2126.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+bytes32 msgHash = keccak256(abi.encodePacked(ETH_PERSONAL_MESSAGE, hash));
+```
+
+`v`, `r`, and `s` are encoded in the signature byte array using the same scheme as [EIP712 signatures](#EIP712).
+
+### Wallet
+
+The `Wallet` signature type allows an EOA account to sign messages on behalf of a smart contract account.
+
+As per [EIP-1271](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1271.md), smart contract accounts can validate signatures using the public method `isValidSignature`. This method has the following interface:
```
Comment (axic) [inline EIPS/eip-2126.md]: Same here with the URL.

**Comment by axic** [comment]
> Inspired by 0x Protocol V2.0 Signature Types

You can add an Acknowledgement section before Copyright stating this.

File: EIPS/eip-2126.md (hunk 4 above)
Comment (hiddentao) [reply EIPS/eip-2126.md]: Would renaming the `Motivation` section to `Rationale` suffice?

**Comment by hiddentao** [comment]
@axic I'd like to be able to commit changes to this PR? could I get access please.

**Comment by axic** [comment]
@hiddentao you need to ask @pedrouid to give you access on his fork.

**Comment by pedrouid** [comment]
Just added you @hiddentao 

https://github.com/pedrouid/EIPs/invitations

**Comment by hiddentao** [comment]
Awesome, thanks!

-- 
Ramesh Nair

Director
HiddenTao Ltd (UK company no. 6807289)

http://hiddentao.com
https://github.com/hiddentao
https://www.linkedin.com/in/hiddentao


On 8 August 2019 at 20:26:20, Pedro Gomes (notifications@github.com) wrote:

Just added you @hiddentao

https://github.com/pedrouid/EIPs/invitations

**Comment by hiddentao** [comment]
@axic Addressed the issues.

File: EIPS/eip-2126.md [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+**EIP712**
+
+An [`EIP712`](https://eips.ethereum.org/EIPS/eip-712) signature is considered valid if the address recovered from calling ecrecover with the given hash and decoded `v`, `r`, `s` values is the same as the specified signer. In this case, the signature is encoded in the following way:
+
+| Offset | Length | Contents            |
+| ------ | ------ | ------------------- |
+| 0x00   | 1      | v (always 27 or 28) |
+| 0x01   | 32     | r                   |
+| 0x21   | 32     | s                   |
```
Comment (Amxx) [inline EIPS/eip-2126.md]: Usually wallet encode it r,s,v (see my comment https://ethereum-magicians.org/t/erc-2126-signature-type-recognition/3392/13?u=amxx)

**Comment by axic** [comment]
@pedrouid @hiddentao let me know if this is ready for merging as draft

**Pull request closed without merging**